gtsam/linear/AcceleratedPowerMethod.h

```diff
@@ -0,0 +1,174 @@
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file   AcceleratedPowerMethod.h
+ * @date   Sept 2020
+ * @author Jing Wu
+ * @brief  accelerated power method for fast eigenvalue and eigenvector
+ * computation
+ */
+
+#pragma once
+
+#include <gtsam/linear/PowerMethod.h>
+
+namespace gtsam {
+
+using Sparse = Eigen::SparseMatrix<double>;
+
+/**
+ * \brief Compute maximum Eigenpair with accelerated power method
+ *
+ * References :
+ * 1) Rosen, D. and Carlone, L., 2017, September. Computational
+ * enhancements for certifiably correct SLAM. In Proceedings of the
+ * International Conference on Intelligent Robots and Systems.
+ * 2) Yulun Tian and Kasra Khosoussi and David M. Rosen and Jonathan P. How,
+ * 2020, Aug, Distributed Certifiably Correct Pose-Graph Optimization, Arxiv
+ * 3) C. de Sa, B. He, I. Mitliagkas, C. Ré, and P. Xu, “Accelerated
+ * stochastic power iteration,” in Proc. Mach. Learn. Res., no. 84, 2018, pp.
+ * 58–67
+ *
+ * It performs the following iteration: \f$ x_{k+1} = A * x_k - \beta *
+ * x_{k-1} \f$ where A is the aim matrix we want to get eigenpair of, x is the
+ * Ritz vector
+ * 
+ * Template argument Operator just needs multiplication operator
+ *
+ */
+template <class Operator>
+class AcceleratedPowerMethod : public PowerMethod<Operator> {
+
+  double beta_ = 0;  // a Polyak momentum term
+
+  Vector previousVector_;  // store previous vector
+
+ public:
+  /**
+   * Constructor from aim matrix A (given as Matrix or Sparse), optional intial
+   * vector as ritzVector
+   */
+  explicit AcceleratedPowerMethod(
+      const Operator &A, const boost::optional<Vector> initial = boost::none,
+      double initialBeta = 0.0)
+      : PowerMethod<Operator>(A, initial) {
+    // initialize Ritz eigen vector and previous vector
+    this->ritzVector_ = initial ? initial.get() : Vector::Random(this->dim_);
+    this->ritzVector_.normalize();
+    previousVector_ = Vector::Zero(this->dim_);
+
+    // initialize beta_
+    if (!initialBeta) {
+      beta_ = estimateBeta();
+    } else
+      beta_ = initialBeta;
+  }
+
+  /**
+   * Run accelerated power iteration to get ritzVector with beta and previous
+   * two ritzVector x0 and x00, and return y = (A * x0 - \beta * x00) / || A * x0
+   * - \beta * x00 ||
+   */
+  Vector acceleratedPowerIteration (const Vector &x1, const Vector &x0,
+                        const double beta) const {
+    Vector y = this->A_ * x1 - beta * x0;
+    y.normalize();
+    return y;
+  }
+
+  /**
+   * Run accelerated power iteration to get ritzVector with beta and previous
+   * two ritzVector x0 and x00, and return y = (A * x0 - \beta * x00) / || A * x0
+   * - \beta * x00 ||
+   */
+  Vector acceleratedPowerIteration () const {
+    Vector y = acceleratedPowerIteration(this->ritzVector_, previousVector_, beta_);
+    return y;
+  }
+
+  /// Tuning the momentum beta using the Best Heavy Ball algorithm in Ref(3)
+  double estimateBeta() const {
```

Comment:
Explain again why this is still a method? I thought David had asked to make this a free function or at least a static method.

Reply:
@dellaert @jingwuOUO I think it's fine to have this as a method (in fact, probably necessary, since we need access to the matrix-vector multiplication operator that's passed in during construction).  What I **don't** want is this function being called inside the constructor (as currently happens in line 71), since this is a heavy-duty operation that is not actually required in order to construct an instance of this object -- permitting a default value of beta = 0.0 is perfectly fine, and I think arguably the "right" default behavior, since that will exactly correspond to the usual (simple) power iteration